### PR TITLE
Added Configurable class for hidden layer controlling

### PR DESCRIPTION
**Pull request by akshitagarwal1998**
This PR adds TransformerEmbeddingCosineConfigurable to be able to add the custom layers and hidden dimension

Usage:

embedding_network:
  _target_: networks.transformer_nn.TransformerEmbeddingCosineConfigurable
  input_features: 1536
  nhead: 12
  num_layers: 6
  dim_feedforward: 3072
  hidden_layer: 1536
  res_block_layers:
   # - [1536, 12]
     - [768, 12]
    #- [364, 12]

### CODE REVIEW

File: src/networks/transformer_nn.py [hunk 1]
```diff
@@ -76,3 +76,93 @@ def forward(self, x, x_mask, y, y_mask):
...
+        else:
+            res_block = OrderedDict()
+            in_dim = input_features
+            block_id = 0  # needed for unique block names
+
+            for out_dim, count in res_block_layers:
+                for i in range(count):
+                    res_block[f'block{block_id}'] = ResBlock(in_dim, out_dim, self.dropout)
+                    in_dim = out_dim  # update input for next block
+                    block_id += 1
```
Comment (bioinsilico) [inline src/networks/transformer_nn.py]: No need to add an index logic, use [enumerate](https://www.geeksforgeeks.org/enumerate-in-python/)

File: src/networks/transformer_nn.py [hunk 2]
```diff
@@ -76,3 +76,93 @@ def forward(self, x, x_mask, y, y_mask):
...
+        )
+        self.transformer = nn.TransformerEncoder(encoder_layer, num_layers=num_layers)
+        if res_block_layers == 0:
+            self.embedding = nn.Sequential(OrderedDict([
+                ('norm', nn.LayerNorm(input_features)),
+                ('dropout', nn.Dropout(p=self.dropout)),
+                ('linear', nn.Linear(input_features, hidden_layer)),
+                ('activation', nn.ReLU())
+            ]))
+        else:
```
Comment (bioinsilico) [inline src/networks/transformer_nn.py:56]: Add the residual logic to the network class `TransformerEmbeddingCosine` and remove `TransformerEmbeddingCosineConfigurable`

**Review by bioinsilico — CHANGES_REQUESTED** [review]
Thank you, @akshitagarwal1998. This is a good improvement. Some minor adjustments.

File: src/networks/transformer_nn.py (hunk 1 above)
Comment (akshitagarwal1998) [reply src/networks/transformer_nn.py]: Resolved.

File: src/networks/transformer_nn.py (hunk 2 above)
Comment (akshitagarwal1998) [reply src/networks/transformer_nn.py:56]: Merged.

**Comment by akshitagarwal1998** [comment]
I have updated the schema to add list of hidden params and updated the TransformerEmbeddingCosine Class.

File: src/networks/transformer_nn.py [hunk 3]
```diff
@@ -54,16 +53,22 @@ def __init__(
...
             ]))
         else:
-            res_block = OrderedDict([(
-                f'block{i}',
-                ResBlock(input_features, hidden_layer, self.dropout)
-            ) for i in range(res_block_layers)])
+            res_block = OrderedDict()
+            in_dim = input_features
+
+            for out_dim, count in res_block_layers:
```
Comment (bioinsilico) [inline src/networks/transformer_nn.py]: This should look something like:

```python
for idx, (out_dim, count) in enumerate(res_block_layers):
```

File: src/config/schema_config.py [hunk 4]
```diff
@@ -41,6 +41,7 @@ class NetworkParams:
     num_layers: int = MISSING
     dim_feedforward: int = MISSING
     hidden_layer: int = MISSING
+    res_block_layers: List[List[int]] = MISSING
```
Comment (bioinsilico) [inline src/config/schema_config.py:44]: You can be more accurate here. Check the syntax but it should be something as:
```python
List[tuple(int,int)] = MISSING
```

File: src/training/embedding_tm_score.py [hunk 5]
```diff
@@ -19,6 +19,9 @@
 from config.schema_config import TrainingConfig
 from lightning.pytorch.loggers import TensorBoardLogger
 
+# Use of high precision for matrix multiplication
+import torch
+torch.set_float32_matmul_precision('high')
```
Comment (bioinsilico) [inline src/training/embedding_tm_score.py:24]: To be in the safe sport comment this line. We will run some training using low.

**Review by bioinsilico — CHANGES_REQUESTED** [review]
Thank you, @akshitagarwal1998. Some adjustments are required.

File: src/training/embedding_tm_score.py (hunk 5 above)
Comment (akshitagarwal1998) [reply src/training/embedding_tm_score.py:24]: So, the only options that we have are ['medium', 'high' or 'highest'(this is the default)].

File: src/config/schema_config.py (hunk 4 above)
Comment (akshitagarwal1998) [reply src/config/schema_config.py:44]: Support missing from Hydra.

File: src/networks/transformer_nn.py (hunk 3 above)
Comment (akshitagarwal1998) [reply src/networks/transformer_nn.py]: for idx, (out_dim, count) in enumerate(res_block_layers):
                for _ in range(count):
                    res_block[f'block{block_idx}'] = ResBlock(in_dim, out_dim, self.dropout)
                    in_dim = out_dim
                    block_idx += 1

Comment (bioinsilico) [reply src/networks/transformer_nn.py]: Something doesn't feel right here. Why not something like:
```python
        for i_idx, (out_dim, count) in enumerate(res_block_layers):
            for j_idx in range(count):
                res_block[f'block_{i_idx}_{j_idx}'] = ResBlock(in_dim, out_dim, self.dropout)
                in_dim = out_dim
```

File: src/networks/transformer_nn.py (hunk 2 above)
Comment (bioinsilico) [reply src/networks/transformer_nn.py:56]: Was this merged? Aren't both classes still available?

**Review by bioinsilico — CHANGES_REQUESTED** [review]
Adjustments needed
